Windows 10 Calculator: resolve expression entry repetitions

josephsl opened the pull request:
### Link to issue number:
Fixes #9428

### Summary of the issue:
Calculator raises UIA notification event (live region change event in older releases and in some screens) to communicate progress and results of calculations. However, this results in repetitions if typed character announcement is turned on. Thus, as part of this new app module, suppress repetitive announcements if possible.

### Description of how this pull request fixes the issue:
This new app module performs the following:

1. Detects conditions for live region change, name change, and UIA notification events that will suppress NVDA from announcing expressions.
2. Adds a script for Enter (regular and numpad) and Escape keys that will allow NVDA to announce current result.

### Testing performed:
Tested with Windows 10 App Essentials implementation (identical to this pull request).

### Known issues with pull request:
None

### Change log entry:
Bug fixes: In Windows 10 Calculator, NVDA will not announce progress of calculations if speak typed characters is disabled. (#9428)

### Additional notes:
Because Windows 10 Calculator is open-source now, changes from inside and outside Microsoft will impact NVDA's support for this app, thus at least one person must be in charge of updating the app module for this app on a regular basis. Also, the new app module is a donation from Windows 10 App Essentials add-on, so it went through years of testing and real-life experiences.

Thanks.

josephsl committed:
Windows 10 Calculator: resolve expression entry repetitions. Re #9428.

Calculator raises UIA notification event (live region change event in older releases and in some screens) to communication progress and results of calculations. However, this results in repetitoins if typed character announcement is turned on. Thus, as part of this new app module, suppress repetitive announcements if possible.

josephsl committed:
Calculator: remove heading level announcement workaround from Windows 10 App Essentials (accidentally committed, now gone).

leonardder commented on source/appModules/calculator.py:
```diff
@@ -0,0 +1,88 @@
+
+	_shouldAnnounceResult = False
+	# Name change says the same thing multiple times for some items.
+	_resultsCache = ""
```
This variable name suggests that it stores a cache of multiple results, which it doesn't.

```suggestion
	_cachedResult = ""
```

lukaszgo1 replied:
@josephsl I don't think you've addressed this comment.

leonardder commented on source/appModules/calculator.py:
```diff
@@ -0,0 +1,88 @@
+		if isinstance(obj, UIA) and obj.UIAElement.cachedAutomationID != "CalculatorResults" and obj.name != self._resultsCache:
+			# For unit conversion, UIA notification event presents much better messages.
+			if obj.UIAElement.cachedAutomationID not in ("Value1", "Value2"):
+				ui.message(obj.name)
```
Why not call super here? Isn't this the focused object?

leonardder commented on source/appModules/calculator.py:
```diff
@@ -0,0 +1,88 @@
+	def event_liveRegionChange(self, obj, nextHandler):
+		# Unfortunately, the control that fires this has no automation ID yet says it is a generic text block.
+		# This may mean anything can be announced, so try to filter them.
+		if shouldLiveRegionChangeProceed(obj):
```
Have you considered placing an overlay class on these objects that overrides `_shouldAllowUIALiveRegionChangeEvent`?

leonardder commented on source/appModules/calculator.py:
```diff
@@ -0,0 +1,88 @@
+		try:
+			shouldAnnounceNotification = obj.previous.UIAElement.cachedAutomationID in ("numberPad", "UnitConverterRootGrid")
+		except AttributeError:
+			shouldAnnounceNotification = api.getForegroundObject().children[1].lastChild.firstChild.UIAElement.cachedAutomationID != "CalculatorResults"
```
I would much prefer catching exceptions for this one as well, just to make sure that future changes won't break this in a major way.

leonardder commented on source/appModules/calculator.py:
```diff
@@ -0,0 +1,88 @@
+			if focus.UIAElement.cachedAutomationID == "CalculatorResults":
+				queueHandler.queueFunction(queueHandler.eventQueue, focus.reportFocus)
+			else:
+				resultsScreen = api.getForegroundObject().children[1].lastChild
```
Could you also catch exceptions for this one?

leonardder commented on source/appModules/calculator.py:
```diff
@@ -0,0 +1,88 @@
+					queueHandler.queueFunction(queueHandler.eventQueue, resultsScreen.firstChild.reportFocus)
+
+	# Without this, gesture binding fails even with script decorator deployed.
+	__gestures={}
```
Now this is interesting. Have you looked at the possible cause of this? If not, it might help to file a separate issue for this. The script decorator code is pretty scary.

josephsl commented:
Hi,

Will plan to experiment with it (likely tonight provided my health permits).

Thanks for reminding me about that flag.

josephsl commented:
Hi,

Regarding script decorator issue, I'm thinking it has to do with name mangling.

Overall work: I'm willing to delay this work until Python 3 comes to NVDA. Doing so could also give me more time to experiment with things here and there.

Thanks.

josephsl commented:
Hi,

Regarding live region change event: after looking at the UIA live region flag and my own code, it might be possible, but we run into an important design limitation: when @MIchaelDCurran designed live region change handling for UIA objects in 2017 (with revisions afterwards), it was intended that all live regions will be announced, provided that texts are different and there is a reasonable time interval between event firings. Part of the reason for writing that function in the Calculator app module was to detect which regions should be announced, because it isn't quite possible to name all live region descriptors in one go and lump them into an overlay class. Perhaps the newly opened Calculator will change that, but that'll take months to hunt and fix. Besides, detecting live regions and defining the regions to be silenced will introduce code duplication like so:

1. Descriptors of live regions in chooseNVDAObjectOverlayClasses.
2. Flagging which live regions to announce in obj._get__shouldAllowUIALiveRegionChangeEvent.

Both will describe the same thing. Resolving the underlying issue that causes this code duplication will involve opening up other possibilities to silence live regions.

Thanks.

josephsl commented:
Hi,

If talking about the result, it isn't necessarily a focused object because it'll be announced as soon as enter key is pressed to finalize the expression.

Thanks.

josephsl commented:
Hi,

Update: after doing more tests, it appears @LeonardDer's intuition was correct - disabling live region announcement will do. But first, I'll let people test possible changes via my add-on before editing the pull request.

Thanks.

josephsl committed:
Merge branch 'master' into win10calculator

josephsl committed:
Windows 10 Calculator: remove live region handler.

In the old days of Calculator, live regions were used from various places, including calculator expressions, unit conversion and other places. Now that uIA notification event and some name change events will take care of this, remove live region change event support code (and also becasue NVDA itself will announce live regions if possible).

josephsl committed:
Calculator: updates to notification and name change event handling due to UI changes in 2019.

josephsl commented:
Hi,

Two things:

1. Actually, name change event handling was modified - we still need to keep live regions alive for date calculation at least.
2. Flake8 gives all sorts of problems with indents. If I fix E121 (visual appearance), it comes back with E101 (mixed spaces and indents). I think it would be best to ignore visual appearance bit.

Thanks.

josephsl commented:
Hi,

Actually, ET128.

josephsl committed:
Merge branch 'master' into win10calculator

josephsl committed:
Calculator app module: lint fixes.

lint results include splitting each part of a conditional, spaces between assignment operator, and block comments.

josephsl commented:
Hi,

Calculator has gone through yet another UI redesign, this time due to "always on" i.e. picture-in-picture mode support. The redesign also rearranges the UIA tree, which breaks this PR and the app module that ships with Windows 10 App Essentials. I'll prepare a fix after testing this week's Insider Preview build (if any).

Thanks.

josephsl committed:
Merge branch 'master' into win10calculator

josephsl committed:
Calculator: gather a list of UIA automation ID's for objects that should not allow NVDA to repeatedly announce calculation results or expressions. Re #9428.

With the advent of alwasy on top i.e. compact overlay mode in Calculator, additional UIA automation ID's were added for calculation results and expressions in overlay mode. Because of ine length concerns, the additional ID's, together with the existing ones, are now housed inside a list of known ID's causing NVDA to announce expressions repeatedly. Not only this makes it easy to add more ID's in the future (either through NVDA Core or add-ons), it improves readability for name change and notification event handler methods (and keeps the lines short).

josephsl committed:
Calculator: support always on top (compact overlay) mode.

In August 2019, Calculator app (preview) added always on top mode. As the name suggests, this puts standard calculator on top of other windows. Because of this, UIA tree has been changed, so work with old and new tree structure.

josephsl commented:
To @LeonardDer: ready for another round of review, if you are willing.

Thanks,

josephsl committed:
Merge branch 'master' into win10calculator

josephsl committed:
Calculator: add additional UIA automation ID's and class names to items that should not be announced. Re #9428.

josephsl committed:
Calculator: prevent NVDA from announcing name changes for more controls. Re #9428.

josephsl committed:
Calculator: update copyright year

josephsl committed:
Calculator: remove redundant gestures dictionary due to use of script decorator. Re #9428.

josephsl commented:
Hi,

Update: since the last time we talked, graphing mode has been added to Calculator, but supporting that is beyond the scope of this PR. Giving this PR a bit of break due to backlog (otherwise ready for review).

Thanks.

josephsl commented:
Hi,

Update: it turns out the solution is simpler than I thought: rather than adding individual UIA notification activity ID's, we can just take care of one or two corner cases (DisplayUpdate ID should not be announced unless noted otherwise). This is thanks to the fact that Calculator app source code is hosted on GitHub (MIT license), so we now know the full list of activity ID's.

Thanks.

josephsl committed:
Merge branch 'master' into win10calculator

josephsl committed:
Calculator: style (remove file name at the start of the file)

josephsl committed:
Calculator: remove unneeded comments

josephsl committed:
Calculator: announce more notifications except display updated activity ID.

Display updated activity ID should be ignored unless needed, as it causes repetitions when entering calculations while speak typed characters is on.

josephsl committed:
Merge branch 'master' into win10calculator

josephsl committed:
Windows 10 Calculator: cachedAutomationID -> UIAAutomationId.

feerrenrut commented:
@josephsl There are a number of remaining review actions on this PR. Please can you check them out comment / resolve / make changes as appropriate. Then we can move onto a second review.

josephsl commented:
Hi,

Addressed in subsequent commits - will mark them as such. Thanks for the reminder.

feerrenrut commented:
@leonardder since you reviewed this previously, and the feedback has now all been addressed, could you please review this again.

josephsl commented:
Hi, I think I have, but may need to look into this again. Besides, I do need to replace UIAElement.cachedAutomationId with UIAAutomationId soon. Thanks.

 

From: Łukasz Golonka <notifications@github.com> 
Sent: Sunday, November 15, 2020 2:38 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Windows 10 Calculator: resolve expression entry repetitions (#9429)

 

@lukaszgo1 commented on this pull request.

  _____  

In source/appModules/calculator.py <https://github.com/nvaccess/nvda/pull/9429#discussion_r523824120> :

> +     if automationID == "DateDiffAllUnitsResultLabel":
+              return True
+       elif automationID == "":
+              prevAutomationID = obj.previous.UIAElement.cachedAutomationID
+              if prevAutomationID == "negateButton":
+                      return False
+              elif prevAutomationID == "numberPad":
+                      return api.getForegroundObject().children[1].children[3].UIAElement.cachedAutomationID != "CalculatorResults"
+       return False
+
+
+class AppModule(appModuleHandler.AppModule):
+
+       _shouldAnnounceResult = False
+       # Name change says the same thing multiple times for some items.
+       _resultsCache = ""

@josephsl <https://github.com/josephsl>  I don't think you've addressed this comment.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9429#discussion_r523824120> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEDHTF6PQHMO3S3KG43SQBJ3DANCNFSM4HCTR7TQ> .

josephsl committed:
Merge branch 'master' into win10calculator

LeonarddeR reviewed (dismissed):
@feerrenrut sorry for missing the request for review.

josephsl commented:
Hi, note that notification event is fired when working with graphing calculator mode, but given the projection that it won’t be used a lot, I’m not going to revise this PR for sake of that. Windows 10 App Essentials add-on version will continue to include experimental support for graphing calculator UI for now until it is ready for Core at a later date. Thanks.

feerrenrut reviewed (dismissed):
Thanks @josephsl

feerrenrut committed:
Update changes file for PR #9429

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into win10calculator

josephsl commented:
Hi,

Two issues found, to be resolved in a PR (coming today):

1. Results are not announced in compact overlay (always on top) mode. This will require tweaking result announcement script to include a second UIA Automation Id. A much better solution is locating and announcing results element (next year because I would like to test this solution with users through WinTenApps add-on).
2. In unit converter, source and target values (and units) are announced twice if focused on these fields, caused by notification and name change events being fired one after another. This can be resolved by adding two new Automation Id's to a list of Id's that should be checked for suppressing announcements.

Thanks.
